### DISCUSSION TITLE
Prevent flapping between two states

### DISCUSSION THREAD

**Discussion by kad3nce** [statelyai/xstate#1879] [Q&A]
I'm using xstate to model a network monitor. When the state machine is in the OFFLINE state, I will show a warning banner to the user, so I need to prevent flapping between ONLINE/OFFLINE states. How can I, for example, enforce that any time the OFFLINE state is entered, it will stay in that state for at least 10 seconds?

**Comment by ezhikov** [comment]
Depending on your requirements there can be different ways to achieve this. First approach is very straightforward, but not very flexible. You add internal state into `OFFLINE`. Let's call these internal states "accepting_messages" and "not_accepting_messages" and make latter initial. Then you add [delayed transition](https://xstate.js.org/docs/guides/delays.html) from `not_accepting_messages` to `accepting_messages` and there you put transitions to `ONLINE` state. Something like this:

```json5
{
  OFFLINE: {
    initial: "NOT_ACCEPTING_MESSAGES",
    states: {
      NOT_ACCEPTING_MESSAGES: {
        after: {
          10000: "ACCEPTING_MESSAGES",
        },
      },
      ACCEPTING_MESSAGES: {
        // put logic here
      },
    },
  },
  ONLINE: {},
}
```

If you want to accept messages, but want react to them not earlier than in 10 seconds you can add some sort of queue (array is good) into context, and on every message in `NOT_ACCEPTING_STATE` you will fill this query. Then, upon entry into `ACCEPTING_MESSAGES` you can either send or raise all those events (`send` actions creator supports sending array).

```json5
{
  OFFLINE: {
    initial: "NOT_ACCEPTING_MESSAGES",
    states: {
      NOT_ACCEPTING_MESSAGES: {
        on: {
          "*": { actions: "put_message_into_queue" },
        },
        after: {
          10000: "ACCEPTING_MESSAGES",
        },
      },
      ACCEPTING_MESSAGES: {
        entry: ["process_queued_messages", "clear_queue"],
        // put logic here
      },
    },
  },
  ONLINE: {},
}
```
And implementation of actions can go something like this:
```js
let put_message_into_queue = assign({queue: (ctx, message) => ctx.queue.concat(event)})
let clear_queue = assign({queue: []});
let process_queued_messages = pure((context) => {
  if (context.queue.length) return [send(context.queue)];
  return [];
})
```

**Reply by davidkpiano** [reply]
Yep, nested states are a great solution for this!

In general, the solution to any state modeling problem can be found by translating your question to states + events:

> How can I, for example, enforce that any time the OFFLINE state is entered, it will stay in that state for at least 10 seconds?

Becomes this:

> How can I enforce that when entering the OFFLINE state, I also enter **an "offline for at least 10 seconds" state**?

And then the answer ("nested offline states") becomes apparent.

**Reply by kad3nce** [reply]
Excellent. Thank you both. I should get a chance to try this out next week. Will report back once I do.

**Reply by kad3nce** [reply]
That worked great. 👍
